### PR TITLE
Topic Event Collector

### PR DESCRIPTION
**Pull request by jeqo**
To maintain metadata about Topics, first Topics need to be collected from a Cluster. 
To do this, a Topic Event Collector will be implemented. 
This Collector will rely on a Kafka Topic to store events. This events will be used to recreate current state and validate events from querying the Kafka Cluster.
Event though this module is not a main part of Kafka Metadata application, it will be kept on this repository until we define another host repository with better naming/responsibility.

### CODE REVIEW

File: collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollector.scala [hunk 1]
```diff
@@ -29,8 +29,8 @@ class TopicCollector(system: ActorSystem) {
     system.actorOf(TopicRepository.props(appConfig.Kafka.bootstrapServers))
   val topicManager: ActorRef =
     system.actorOf(TopicManager.props(appConfig.Collector.pollFrequency, topicRepository, topicEventProducer))
-  val topicEventConsumer: ActorRef =
-    system.actorOf(TopicEventConsumer.props(topicManager))
+  val topicEventConsumer: TopicEventConsumer =
+    TopicEventConsumer(topicManager, appConfig.Collector.topicEventTopic, appConfig.Kafka.bootstrapServers)
```
Comment (zhenik) [inline collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollector.scala]: I disagree with current architecture.
The architecture is flat. [Akka-antipatterns](https://manuel.bernhardt.io/2016/08/09/akka-anti-patterns-flat-actor-hierarchies-or-mixing-business-logic-and-failure-handling/)
It should be parent-child as much as possible. 
Take a look at offer in link [architecture](https://www.lucidchart.com/invitations/accept/39c37d1f-0b4d-4434-9346-4eccf1eab1d0).

**Review by zhenik — COMMENTED** [review]
Yes, I agree. I see this topic-collector as a part of other system at the end.

File: collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollector.scala (hunk 1 above)
Comment (jeqo) [reply collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollector.scala]: Agree with supervision, but your diagram looks pretty much as it is now. Maybe what you are talking about is how/where actors are created right?

File: collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollector.scala [hunk 2]
```diff
@@ -1,8 +1,29 @@
...
+
+  val config: Config = ConfigFactory.load()
+  val appConfig: TopicCollectorConfig = new TopicCollectorConfig(config)
+
+  val topicEventProducer: ActorRef = system.actorOf(TopicEventProducer.props())
+  val topicRepository: ActorRef = system.actorOf(TopicRepository.props(appConfig.Kafka.bootstrapServers))
+  val topicManager: ActorRef = system.actorOf(TopicManager.props(appConfig.Collector.pollFrequency, topicRepository, topicEventProducer))
+  val topicEventConsumer: ActorRef = system.actorOf(TopicEventConsumer.props(topicManager))
+
+  topicManager ! CollectTopics
```
Comment (zhenik) [inline collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollector.scala]: I think, this is not necessary, you can send init-message inside actor to itself on `PostConstruct` phase.

File: collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollector.scala (hunk 1 above)
Comment (zhenik) [reply collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollector.scala]: This diagram contains 2 pages. 1st - current architecture, 2nd - offer

File: collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollector.scala (hunk 2 above)
Comment (jeqo) [reply collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollector.scala]: Good, I will review this as part of the refactor or Actor hierarchy,

File: collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollector.scala (hunk 1 above)
Comment (jeqo) [reply collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollector.scala]: haha, sorry. Not sure the communication on your offer reflects the current logic, but I got the idea. Next push should include some fixes about this.

**Comment by jeqo** [comment]
@NikitaZhevnitskiy take a look to the latest changes. 

> Ideas about next tasks: HTTP Server for data-access and Metrics for Collector.

File: collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollectorConfig.scala [hunk 3]
```diff
@@ -0,0 +1,15 @@
+package no.sysco.middleware.kafka.metadata.collector.topic
+
+import java.time.Duration
+
+import com.typesafe.config.Config
+
+class TopicCollectorConfig(config: Config) {
+  object Collector {
+    val pollInteval: Duration = config.getDuration("collector.topic.poll-interval")
```
Comment (zhenik) [inline collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollectorConfig.scala]: typo: topicInterval

File: collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/internal/TopicEventConsumer.scala [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+  val config: Config = context.system.settings.config.getConfig("akka.kafka.consumer")
+  val consumerSettings: ConsumerSettings[String, Array[Byte]] =
+    ConsumerSettings(config, new StringDeserializer, new ByteArrayDeserializer)
+      .withBootstrapServers(bootstrapServers)
+      .withPollInterval(100 millis)
+      .withProperty(ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false")
+      .withProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
+      .withGroupId("kafka-metadata-collector-topic-internal-consumer")
+
+  val killSwitch: UniqueKillSwitch =
```
Comment (zhenik) [inline collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/internal/TopicEventConsumer.scala:39]: Could u elaborate more about killSwitch, or provide link(reference)?

File: collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollector.scala (hunk 1 above)
Comment (zhenik) [reply collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollector.scala]: It is still quite flat, but u got idea, we will fix it later.

File: collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollectorConfig.scala (hunk 3 above)
Comment (jeqo) [reply collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollectorConfig.scala]: or `topicPollInterval`?

Comment (zhenik) [reply collector/topic/src/main/scala/no/sysco/middleware/kafka/metadata/collector/topic/TopicCollectorConfig.scala]: yes.

**Pull request merged**